pkg/diff/sql_vertex_generator.go

```diff
@@ -0,0 +1,229 @@
+package diff
+
+import (
+	"fmt"
+
+	"github.com/stripe/pg-schema-diff/internal/schema"
+)
+
+type diffType string
+
+const (
+	diffTypeDelete   diffType = "DELETE"
+	diffTypeAddAlter diffType = "ADDALTER"
+)
+
+// partialSQLGraph represents the SQL statements and their dependencies. Every SQL statement is a
+// vertex in the graph. This is different from a graph because a dependency may exist between a node
+// within this part and a node in another part.
+type partialSQLGraph struct {
+	vertices     []sqlVertex
+	dependencies []dependency
+}
+
+func (s *partialSQLGraph) statements() []Statement {
+	var statements []Statement
+	for _, vertex := range s.vertices {
+		statements = append(statements, vertex.statements...)
+	}
+	return statements
+}
+
+func concatPartialGraphs(parts ...partialSQLGraph) partialSQLGraph {
+	var vertices []sqlVertex
+	var dependencies []dependency
+	for _, part := range parts {
+		vertices = append(vertices, part.vertices...)
+		dependencies = append(dependencies, part.dependencies...)
+	}
+	return partialSQLGraph{
+		vertices:     vertices,
+		dependencies: dependencies,
+	}
+}
+
+func graphFromPartials(parts partialSQLGraph) (*sqlGraph, error) {
+	graph := newSqlGraph()
+	for _, vertex := range parts.vertices {
+		// It's possible the node already exists. merge it if it does
+		if graph.HasVertexWithId(vertex.GetId()) {
+			vertex = mergeVertices(graph.GetVertex(vertex.GetId()), vertex)
+		}
+		graph.AddVertex(vertex)
+	}
+
+	for _, dep := range parts.dependencies {
+		sourceVertex := sqlVertex{
+			id:         dep.source,
+			priority:   sqlPriorityUnset,
+			statements: nil,
+		}
+		targetVertex := sqlVertex{
+			id:         dep.target,
+			priority:   sqlPriorityUnset,
+			statements: nil,
+		}
+
+		// To maintain the correctness of the graph, we will add a dummy vertex for the missing dependencies
+		addVertexIfNotExists(graph, sourceVertex)
+		addVertexIfNotExists(graph, targetVertex)
+
+		if err := graph.AddEdge(sourceVertex.GetId(), targetVertex.GetId()); err != nil {
+			return nil, fmt.Errorf("adding edge from %s to %s: %w", sourceVertex.GetId(), targetVertex.GetId(), err)
+		}
+	}
+
+	return graph, nil
+}
+
+func mergeVertices(old, new sqlVertex) sqlVertex {
+	priority := old.priority
+	if old.priority == sqlPriorityUnset {
+		priority = new.priority
+	}
```

Comment:
thought: Take the highest priority instead of taking an arbitrary one based on order of vertexes?

Reply:
Yes that makes sense! This just maintains the old behavior! "highest probability excluding the unset probability", i.e., `max(lowestProbablity, unset) -> 0`, which is not what we want.